fix: make sure Minestom services stop like Bukkit services

GiantTreeLP opened the pull request:
<!-- 
Thanks for taking your time and creating a pull request. Please note that we will not merge pull requests
which are not following our code style (https://google.github.io/styleguide/javaguide.html). Most of these
rules are checked while compile using checkstyle. On the other hand, please cover relevant code with tests.
These are showing the maintainers what to expect from your pull requests and ensures that changes to your
code will be consistent over time. See for example https://betterprogramming.pub/13-tips-for-writing-useful-unit-tests-ca20706b5368
if you need a bit of guidance while writing your tests.
-->

### Motivation
<!-- Explain the context and why you're making the change (what is the problem solved by this pr) -->

Minestom is a very lightweight framework for implementing specialized Minecraft servers.  
Especially when the full implementation or behavior of vanilla Minecraft is not needed, Minestom offers a way to run services with very little resources.

One issue that arised in conjunction with CloudNet is that services implemented using Minestom would not properly stop. Minestom would shut down and the service would get unregistered from CloudNet but the processes would be kept running.

### Modification
<!-- Describe the modification you've done to the codebase -->

This pull request mimics the behavior of vanilla implementations of the Minecraft Server which all call `System.exit(0);` when their process is done. This change makes sure that the JVM shuts down properly, even if other threads might block an ordinary shutdown from happening (ie. Netty and other networking related tasks).

### Result
<!-- Describe the result of the pull request (what changed compared to before) -->

Services using the Minestom framework now properly shut down and allow CloudNet to advance to the next lifecycle steps.

##### Other context
<!-- Other context of the pull request, a discussion, issue or anything else related -->
Fixes #1304

GiantTreeLP committed:
feat: Make sure Minestom services stop like Bukkit services

This change implements a class transformer for net.minestom.server.MinecraftServer to call System.exit when MinecraftServer.stopCleanly is called.

GiantTreeLP commented:
As of now, this solution also works on server implementations build on hollow-cube/minestom-ce.

0utplay commented:
Can we maybe implement this as discussed here (https://discord.com/channels/713041279277203457/777861579659411466/1182967379488161872)

GiantTreeLP commented:
> Can we maybe implement this as discussed here (https://discord.com/channels/713041279277203457/777861579659411466/1182967379488161872)

I'm going to update my implementation to match what we discussed back then.

GiantTreeLP committed:
feat: Make sure Minestom services stop like Bukkit services

This change implements a class transformer for net.minestom.server.ServerProcessImpl to call System.exit when ServerProcessImpl.stop is called.
The changes also take care of correctly patching the opcode, if label nodes follow the last return statement.

GiantTreeLP commented:
@0utplay I have updated the PR but can't initiate the build required.

Feel free to also add a review to the code, the changes are pretty minimal.

0utplay commented:
Thats because there are conflicts in the Wrapper class. I can resolve them later.

0utplay committed:
Merge branch 'refs/heads/nightly' into fork/1304-minestom-does-not-shut-down-transformer

# Conflicts:
#	wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/Wrapper.java

0utplay commented on wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java:
```diff
@@ -0,0 +1,69 @@
+   * @param method The method to search in
+   * @return The last return instruction in the method, or null if no return instruction was found
+   */
+  private static AbstractInsnNode findLastReturn(MethodNode method) {
```
Can we please add a @NonNull annotation here (return type and param) and document the thrown NullPointerException if the passed method is null

GiantTreeLP replied:
There is no NullPointerException being thrown, I'll change the code to reflect that.

0utplay commented on wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java:
```diff
@@ -0,0 +1,69 @@
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+
+public class MinestomStopCleanlyTransformer implements Transformer {
```
Make this class final please

0utplay commented on wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java:
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019-2023 CloudNetService team & contributors
```
You'll have to run spotless once. (or just change this manually to 2024)

GiantTreeLP committed:
feat: Refactor Minestom transformer

This change improves the quality of the MinestomStopCleanlyTransformer by making it final and handling the unlikely case, that there is no return statement in the method to be patched.

0utplay committed:
chore: small code style nits
